### PR TITLE
Adds feature to show errors and warnings when node is collapsed

### PR DESCRIPTION
**Pull request by netonjm**
Adds feature to show errors and warnings when node is collapsed
Fixes #570136
Fixes #570560

image

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs [hunk 1]
```diff
@@ -228,6 +232,21 @@ public void AddNode (BuildOutputNode node)
 			get => nodes.Any (x => x.HasData);
 			set => throw new NotImplementedException ();
 		}
+
+		public override int Warnings {
+			get => nodes.First ()?.Warnings ?? 0;
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: What you want is `FirstOrDefault` here.

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: Same below.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs [hunk 2]
```diff
@@ -70,6 +70,10 @@ class BuildOutputNode : TreePosition
 		public virtual string Project { get; set; }
 		public virtual int LineNumber { get; set; }
 
+		public virtual bool ShowsInformation { get; set; }
+		public virtual int Errors { get; set; }
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: I'm not sure I get what this `ShowsInformation` is used for, is it to check if this node should display the errors/warnings counts when collapsed? If so, this doesn't belong to the `BuildOutputNode`, i.e. the data model, but to the cell renderer, the view.

**Comment by rodrmoya** [comment]
We only display the errors/warnings indicator if there are errors/warnings, so looks a bit weird when nodes have only errors or warnings, but not both:

screen shot 2018-02-23 at 15 21 57

Should we always display both even when there are 0 errors or 0 warnings?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs (hunk 1 above)
Comment (netonjm) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: ahh yes :)

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs (hunk 2 above)
Comment (rodrmoya) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: Also, I think these should be named ErrorsCount and WarningsCount?

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs [hunk 3]
```diff
@@ -225,6 +227,16 @@ public void AddNode (BuildOutputNode node)
...
+		public override int WarningCount {
+			get => nodes.FirstOrDefault ()?.WarningCount ?? 0;
+			set => throw new NotImplementedException ();
+		}
+
+		public override int ErrorCount {
+			get => nodes.FirstOrDefault ()?.ErrorCount ?? 0;
+			set => throw new NotImplementedException ();
+		}
 	}
```
Comment (rodrmoya) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.BuildOutputView/BuildOutputNode.cs]: this doesn't affect this branch, as the icons are only shown for children, but the implementation of these 2 is wrong, it should be the sum of the WarningCount/ErrorCount of all the nodes, not just of the 1st node

**Pull request merged**